source/reduce/remove_selection_reduction_opportunity.cpp

```diff
@@ -0,0 +1,73 @@
+// Copyright (c) 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/basic_block.h"
+#include "source/opt/ir_context.h"
+
+#include "remove_selection_reduction_opportunity.h"
+
+namespace spvtools {
+namespace reduce {
+
+bool RemoveSelectionReductionOpportunity::PreconditionHolds() {
+  // It is possible that another opportunity may have changed the block so that
+  // it no longer heads a selection construct.
+  return block_->GetMergeInst() &&
+         block_->GetMergeInst()->opcode() == SpvOpSelectionMerge &&
+         block_->terminator()->opcode() == SpvOpBranchConditional;
+}
+
+void RemoveSelectionReductionOpportunity::Apply() {
+  auto merge_instruction = block_->GetMergeInst();
+  auto merge_block_id = merge_instruction->GetSingleWordInOperand(0);
+  auto terminator = block_->terminator();
+
+  terminator->SetOpcode(SpvOpBranch);
+  // Pick one of the former conditional branch operands depending on whether we
+  // are choosing the LHS or RHS.
+  terminator->SetInOperands({choose_lhs_ ? terminator->GetInOperand(1)
+                                         : terminator->GetInOperand(2)});
+
+  // Go through the function looking for breaks out of the selection, i.e.
+  // conditional branches that can jump to the selection's merge block.  Rewrite
+  // these as unconditional branches.
+  for (auto& block : *function_) {
+    for (auto& inst : block) {
+      if (inst.opcode() != SpvOpBranchConditional) {
```

Comment:
I doubt it would be common, but you could have an OpSwitch with a default and one other target (one of which is the selection's merge).

Reply:
I don't get it, sorry.  Would you be able to sketch an example?

Reply:
```
OpSwitch %value %merge 1 %other
; or
OpSwitch %value %other 1 %merge
```

It really just an odd expression of a conditional branch, but I could also do things like:
```
OpSwitch %value %other 1 %merge 2 %merge 3 %merge
```

As long as there is only a single non-merge target I think it is valid to not have a selection merge precede it.

Reply:
I suppose it might be better to look at the predecessors of the merge block; if a predecessor has more than one successor, then I think we need to consider it. The header might be a special case. There might be other issues.

On realizing the subtlety of these examples, it feels tempting to reconsider the previously discussed idea where we instead just try to simplify the conditional branch so both targets are the same, then eliminate dead code. The problematic branches inside the selection will still be valid, but they should then also get simplified in the same way, until the original selection construct can be trivially removed (we can check the predecessors of the merge block perhaps). 

We might reconsider limiting these reduction passes to OpSelectionMerge; we might be able to just consider the conditional branch instructions. Need to consider more examples though.

Another worry: what if we have a continue construct that contains a multi-way branch and we end up "simplifying" this so that the back-edge becomes unreachable; I think this will be invalid. On the other hand, if the validator can detect this it almost doesn't matter.